sympy/polys/matrices/sdm.py

```diff
@@ -970,3 +1012,35 @@ def sdm_particular_from_rref(A, ncols, pivots):
         if Ain is not None:
             P[j] = Ain / A[i][j]
     return P
+
+
+def sdm_applyfunc(M, shape, f, domain):
+    nrows, ncols = shape
+    Mf = {}
+    for i in range(nrows):
+        Mf[i] = {}
+        for j in range(ncols):
+            try:
+                domain_elem = domain.convert(f(M[i][j]))
+                if not domain.is_zero(domain_elem):
+                    Mf[i][j] = domain_elem
+            except KeyError:
+                domain_elem = domain.convert(f(domain.zero))
+                if not domain.is_zero(domain_elem):
+                    Mf[i][j] = domain_elem
+
+        if not bool(Mf[i]):
+            del Mf[i]
+    return Mf
+
+
+def sdm_applyfunc_nonzero(M, f, domain):
+    Mf = {}
+    for i, Mi in M.items():
+        for j, Mij in Mi.items():
+            try:
+                Mf[j][i] = domain.convert(f(Mij))
+            except KeyError:
+                Mf[j] = {i: domain.convert(f(Mij))}
```

Comment:
Why would KeyError ever be raised?

This is just:
```python
Mf = {i: {j: f(Mij) for j, Mij in Mi.items()} for i, Mi in M.items()}
```
Possibly it can be faster (if `f` is a fast function) to use `zip` and `map` like:
```python
Mf = {i: dict(zip(Mi, map(f, Mi.values))) for i, Mi in M.items()}
```